experimental/list_formatter/src/deduplicating_array.rs

```diff
@@ -0,0 +1,253 @@
+        {
+            match r {
+                HumanDe::Value(v) => {
+                    array[i].write(v);
+                }
+                HumanDe::Fallback([j]) => unsafe {
+                    // Fallbacks should always be to a previous value,
+                    // which makes the assume_init_ref safe
+                    if j >= i {
+                        return Err(D::Error::custom(alloc::format!(
+                            "Illegal forward fallback {}->{}",
+                            i,
+                            j
+                        )));
+                    }
+                    array[i].write(array[j].assume_init_ref().clone());
+                },
+            }
+        }
+    } else {
+        for (i, r) in
+            IntoIterator::into_iter(<[MachineDe<T>; N]>::deserialize(deserializer)?).enumerate()
+        {
+            match r {
+                MachineDe::Value(v) => {
+                    array[i].write(v);
+                }
+                MachineDe::Fallback(j) => unsafe {
+                    // Fallbacks should always be to a previous value,
+                    // which makes the assume_init_ref safe
+                    if j >= i {
+                        return Err(D::Error::custom(alloc::format!(
+                            "Illegal forward fallback {}->{}",
+                            i,
+                            j
+                        )));
+                    }
+                    array[i].write(array[j].assume_init_ref().clone());
+                },
+            }
+        }
+    }
+
+    // https://github.com/rust-lang/rust/issues/61956
+    Ok(unsafe { core::ptr::read(array.as_ptr() as *const [T; N]) })
+}
+
+#[derive(serde::Serialize)]
+#[serde(untagged)]
+enum HumanSer<'a, T> {
+    Value(&'a T),
+    Fallback([usize; 1]),
+}
+
+enum MachineSer<'a, T> {
+    Value(&'a T),
+    Fallback(usize),
+}
+
+// These enums are public because they are exposed in the trait bounds of
+// dedupe_array_de. Serde only implements Deserialize<'de> on arrays up to
+// size 32, so [Dedupe<T>;N]: Deserialize<'de> cannot be inferred from
+// Dedupe<T>: Deserialize<'de> in the general case. See
+// https://github.com/serde-rs/serde/issues/1937.
+//
+// These are not considered part of the stable API.
+
+#[derive(serde::Deserialize)]
+#[serde(untagged)]
+#[doc(hidden)]
+pub enum HumanDe<T> {
+    Value(T),
+    Fallback([usize; 1]),
+}
+
+#[doc(hidden)]
+pub enum MachineDe<T> {
+    Value(T),
+    Fallback(usize),
+}
+
+impl<'a, T> Serialize for MachineSer<'a, T>
+where
+    T: Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+        T: Serialize,
+    {
+        match self {
+            // Serialize values as an enum variant with index 0
+            MachineSer::Value(t) => serializer.serialize_newtype_variant(
+                "unused-enum-name",
+                0,
+                "unused-enum-variant",
```

Comment:
Note: You should fill in these strings in order to enable JSON serialization of MachineSer to work.

Reply:
As explained above, it's not possible to fill in the variant string.